src/oop/object_interactions/medium/q4/MainTest.java

```diff
@@ -0,0 +1,192 @@
+        }
+    }
+
+    private static Stream<Arguments> inputProvider() {
+        return Stream.of(
+                Arguments.of('A', 10.3, 'B', 4.9),
+                Arguments.of('A', 10.3, 'A', 7.2),
+                Arguments.of('A', 10.3, 'A', minBlood),
+                Arguments.of('A', 10.3, 'A', minBlood - 1),
+                Arguments.of('A', 10.3, 'A', minBlood + 1),
+                Arguments.of('B', 4.2, 'A', 2.5),
+                Arguments.of('O', 3.4, 'B', 5.7)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void correctTransfusionFromMethod(char type, double volume, char donType, double donVolume) throws Throwable {
+        Object[][] arguments = {
+                {type, char.class},
+                {volume, double.class}
+        };
+        Snail donor = new Snail(donType, donVolume);
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] methodArguments = {
+                {donor, Snail.class}
+        };
+        Object output;
+        if (type != donType) {
+            output = testClass.callMethod(transfusionFromMethodName, methodArguments, new String[]{"public"}, classInstance, new Clause[]{
+                    new StringLiteral("Blood type incompatible"),
+                    new Optional(new StringLiteral(" "))
+            });
+        } else if (donVolume <= minBlood) {
+            output = testClass.callMethod(transfusionFromMethodName, methodArguments, new String[]{"public"}, classInstance, new Clause[]{
+                    new StringLiteral("Blood transfusion starting..."),
+                    new Optional(new StringLiteral(" ")),
+                    new NewLine(),
+                    new StringLiteral("Donor does not have enough blood for a transfusion"),
+                    new Optional(new StringLiteral(" "))
+            });
+        } else {
+            output = testClass.callMethod(transfusionFromMethodName, methodArguments, new String[]{"public"}, classInstance, new Clause[]{
+                    new StringLiteral("Blood transfusion starting..."),
+                    new Optional(new StringLiteral(" ")),
+                    new NewLine(),
+                    new StringLiteral("Blood transfusion completed"),
+                    new Optional(new StringLiteral(" "))
+            });
+            _assertEquals(volume + 1, testClass.getFieldValue(classInstance, volumeAttributeName),
+                    "Your " + transfusionFromMethodName + " method does not correctly add blood to the recipient.");
+            _assertEquals(donVolume - 1, donor.getVolume(),
+                    "Your " + transfusionFromMethodName + " method does not correctly take blood from the donor.");
+
+        }
+        assertNull(output, String.join(" ", "Your", transfusionFromMethodName, "should not return any output"));
+    }
+
+    @ParameterizedTest
```

Comment:
doesn't the question say that the toString is given? I mean, it makes sense that it wouldn't be, but either keep this test and change the question text or get rid of this test and keep the question text unchanged